### DISCUSSION TITLE
Should the main docs contain a section on testing?

### DISCUSSION THREAD

**Discussion by davidharting** [adonisjs/core#3421] [General]
For someone looking for the "Laravel/Rails of Node.js," testing will be top of mind. 
Currently, the only official docs on testing is a [cookbook from June of 2020](https://docs.adonisjs.com/cookbooks/testing-adonisjs-apps).

I think folks would feel more confident adopting Adonis.js if the current state of testing was addressed in the main documentation.  Perhaps in Guides > Fundamentals > Testing. 

The kind of content I would love to see in this page is

* A link to the cookbook on testing (if this is still the recommended path)
* Assurance that baked-in testing will be added to Adonis.js core, and perhaps a timeline for this if there is one
* Some sort of messaging assuring the reader that Adonis.js is very much production-ready and testable. Basically provide a stance and re-assurances on this topic!

My fear is that testing may seem like an after thought for users coming to the current documentation.

**Comment by thetutlage** [comment]
I agree with your broad sentiment, and testing is very important to me.

Along with testing, there are a few more missing pieces in the framework, like support for Queues and rate-limiting, and we will address them all over time. However, to give a rough estimate, we should have first-class support for testing around May 2022.

Now, coming to the importance of testing. You can get some hints from some from the inbuilt APIs around faking an implementation or defining traps. For example:

- [Email traps](https://docs.adonisjs.com/guides/mailer#email-traps) API is designed to trap emails during tests and write assertions on the email content.
- Same is with [Event emitter traps](https://docs.adonisjs.com/guides/events#trapping-events).
- [Global database transactions](https://docs.adonisjs.com/cookbooks/testing-adonisjs-apps#using-global-database-transactions) allows you to wrap all the database operations that occurred during a test inside a transaction and then roll back after each test.
- Similarly, the [Drive fakes](https://docs.adonisjs.com/guides/drive#testing-fake) allows faking the drive instances during tests to avoid uploading files to the configured destination.

It's just a matter of adding these APIs to other modules and then adding a first-class testing engine to the framework. Again, I am fully committed to doing this.

---

> Some sort of messaging assuring the reader that Adonis.js is very much production-ready and testable. Basically provide a stance and re-assurances on this topic!

I am not 100% sure how exactly you want us to convey this. Is it by just writing it as a sentence in the docs or having visible proof with a testing framework and its documentation?

The visible proof will come at its own time. However, putting bold statements without visible proof is something I try to avoid. Also, hard selling is not my skill 🙂

**Reply by davidharting** [reply]
> to give a rough estimate, we should have first-class support for testing around May 2022

Yay! Glad to hear its on the horizon. It appeared to be in active development on Trello, nice to know that means we should look forward to this launching in 2022. 

Thank you very much for pointing me toward areas where the framework is built with testing in mind! It very much feels like a first-class consideration. 

> It's just a matter of adding these APIs to other modules and then adding a first-class testing engine to the framework. Again, I am fully committed to doing this.

This is really the kind of reassurance I personally am looking for, and I hope others who look into Adonis and think about testing also find this thread!

> The visible proof will come at its own time. However, putting bold statements without visible proof is something I try to avoid. 
I love this attitude! You're right. Building the thing is better than a few words in the docs. 

---

I think all of this leaves me with just one further question. Do you expect future major version updates to be done in a similar manner as Adonisjs 5? i.e., Where the major version launches without all modules, and then modules are upgraded in time? 

Or should we expect Adonisjs 6 (whenever that is, even if years from now!) to launch all modules concurrently?

**Reply by thetutlage** [reply]
> I think all of this leaves me with just one further question. Do you expect future major version updates to be done in a similar manner as Adonisjs 5? i.e., Where the major version launches without all modules, and then modules are upgraded in time?

Not at all. `v4 -> v5` was a huge change because we introduced TypeScript and learned from the past mistakes. This is a phase that a lot of frameworks get in their early adoption cycle.

Moving forward all releases will carry forward the existing work. Breaking changes will be subtle with a documented migration path.

With that said. JavaScript ecosystem itself is in the migration phase of CJS to ESM. Once AdonisJS will adopt ESM, then there will be a few breaking changes. I will try my best to keep the breaking changes minimal, but they are not avoidable and if we decide not to go with ESM, then overtime a lot of packages will stop working with AdonisJS.

This is the reality of JavaScript. The language itself is in motion.

**Comment by davidharting** [comment]
Congrats on the March 2022 release of Japa! Such a cool project, and filling a major need in the Node.js ecosystem.
